src/UriNormalize.c

```diff
@@ -245,6 +247,23 @@ static URI_INLINE void URI_FUNC(LowercaseInplace)(const URI_CHAR * first,
 
 
 
+static URI_INLINE void URI_FUNC(LowercaseInplaceAfterPercentEncoding)(const URI_CHAR * first,
+		const URI_CHAR * afterLast) {
+	if ((first != NULL) && (afterLast != NULL) && (afterLast > first)) {
+		URI_CHAR * i = (URI_CHAR *)first;
+		const int lowerUpperDiff = (_UT('a') - _UT('A'));
+		for (; i < afterLast; i++) {
+			if ((*i >= _UT('A')) && (*i <=_UT('Z'))) {
+				*i = (URI_CHAR)(*i + lowerUpperDiff);
+			} else if (*i == _UT('%')) {
+				i += 2;
+			}
```

Comment:
While this seems safe with well-formed data, it could advance the pointer past the end of the array for malformed input. I would rather be safe then sorry. How about:

```suggestion
			} else if (*i == _UT('%')) {
				if (i + 3 >= afterLast) {
					return;
				}
				i += 2;
			}
```